File: go/internal/idle/idle.go
// Package idle implements the Aeron-style spin -> yield -> timed-park backoff
// ladder for the thread-handoff backoff cells. Parameters are fixed constants
// (aeron-go defaults) — cross-language comparability is the point.
//
// Two parkers implement the ladder's park rung:
//
//   - NewBackoff parks with time.Sleep — the idiomatic naive port. Go's
//     runtime timer overshoots sub-millisecond sleeps by orders of magnitude
//     (a 6 µs request costs ~425 µs; >= 8 µs costs ~1 ms — measured in
//     aeron-go 1ce3720), so the ladder collapses to its top rung. That
//     overshoot is deliberately kept: it is what the `backoff` cell measures.
//   - NewYieldingBackoff serves parks shorter than SleepFloorNs by yielding
//     to a deadline (runtime.Gosched), the aeron-go 1ce3720 strategy; parks
//     at or above the floor still sleep. The `backoff_yield` cell.
package idle

import (
	"runtime"
	"time"
)

// Aeron/Agrona BackoffIdleStrategy defaults (aeron-go values).
const (
	MaxSpins  = 10
	MaxYields = 20
	MinParkNs = int64(1000)
	MaxParkNs = int64(time.Millisecond)
	// SleepFloorNs is the shortest park handed to time.Sleep by the yielding
	// variant; shorter parks yield to a deadline instead.
	SleepFloorNs = int64(time.Millisecond)
)

// Backoff walks spin -> yield -> park-doubling on consecutive idle calls and
// resets on work. Single-goroutine use; not safe for concurrent Idle calls.
type Backoff struct {
	spins, yields int
	parkPeriodNs  int64
	// parker performs the park rung; injectable for the ladder tests.
	parker func(ns int64)
}

// NewBackoff returns the naive ladder: every park is a time.Sleep.
func NewBackoff() *Backoff {
	return &Backoff{parkPeriodNs: MinParkNs, parker: sleepPark}
}

// NewYieldingBackoff returns the ladder with sub-floor parks served by
// yielding to a deadline (aeron-go 1ce3720).
func NewYieldingBackoff() *Backoff {
	return &Backoff{parkPeriodNs: MinParkNs, parker: yieldFloorPark}
}

// Idle advances the ladder when workCount == 0 and resets it otherwise.
func (b *Backoff) Idle(workCount int) {
	if workCount > 0 {
		b.spins, b.yields, b.parkPeriodNs = 0, 0, MinParkNs
		return
	}
	switch {
	case b.spins < MaxSpins:
		// Busy rung: count only, matching aeron-go (Go exposes no plain
		// spin hint; procyield is runtime-internal).
		b.spins++
	case b.yields < MaxYields:
		b.yields++
		runtime.Gosched()
	default:
		b.parker(b.parkPeriodNs)
		if next := b.parkPeriodNs * 2; next <= MaxParkNs {
			b.parkPeriodNs = next
		} else {
			b.parkPeriodNs = MaxParkNs
		}
	}
}

func sleepPark(ns int64) {
	time.Sleep(time.Duration(ns))
}

func yieldFloorPark(ns int64) {
	if ns < SleepFloorNs {
		deadline := time.Now().Add(time.Duration(ns))
		for time.Now().Before(deadline) {
			runtime.Gosched()
		}
		return
	}
	time.Sleep(time.Duration(ns))
}

File: rust/thread-handoff/backoff/src/idle.rs
//! Aeron-style spin -> yield -> timed-park backoff ladder (aeron-go default
//! parameters, fixed constants — cross-language comparability is the point).
//! The park rung is `thread::sleep` (nanosleep): Rust's timed sleep
//! overshoots by ~tens of µs on Linux — honest rungs, unlike Go's
//! `time.Sleep` collapse, and that difference is what the grid compares.
//! The parker is injectable so the ladder's state machine is unit-testable
//! without wall-clock assertions (overshoot is host-dependent; it is what
//! the fleet cell measures, not what CI asserts).

use std::thread;
use std::time::Duration;

pub const MAX_SPINS: u32 = 10;
pub const MAX_YIELDS: u32 = 20;
pub const MIN_PARK_NS: u64 = 1_000;
pub const MAX_PARK_NS: u64 = 1_000_000;

/// Backoff ladder over an injectable parker. Single-thread use.
pub struct Backoff<P: FnMut(u64)> {
    spins: u32,
    yields: u32,
    park_ns: u64,
    parker: P,
}

/// The production ladder: parks with `thread::sleep`.
pub fn sleeping() -> Backoff<impl FnMut(u64)> {
    Backoff::with_parker(|ns| thread::sleep(Duration::from_nanos(ns)))
}

impl<P: FnMut(u64)> Backoff<P> {
    pub fn with_parker(parker: P) -> Self {
        Backoff {
            spins: 0,
            yields: 0,
            park_ns: MIN_PARK_NS,
            parker,
        }
    }

    /// Advance the ladder when `work_count == 0`; reset it otherwise.
    pub fn idle(&mut self, work_count: usize) {
        if work_count > 0 {
            self.spins = 0;
            self.yields = 0;
            self.park_ns = MIN_PARK_NS;
            return;
        }
        if self.spins < MAX_SPINS {
            self.spins += 1;
            std::hint::spin_loop();
        } else if self.yields < MAX_YIELDS {
            self.yields += 1;
            thread::yield_now();
        } else {
            (self.parker)(self.park_ns);
            self.park_ns = (self.park_ns * 2).min(MAX_PARK_NS);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;

    fn recorded(parks: &RefCell<Vec<u64>>) -> Backoff<impl FnMut(u64) + '_> {
        Backoff::with_parker(move |ns| parks.borrow_mut().push(ns))
    }

    #[test]
    fn spins_and_yields_before_first_park() {
        let parks = RefCell::new(Vec::new());
        let mut b = recorded(&parks);
        for _ in 0..(MAX_SPINS + MAX_YIELDS) {
            b.idle(0);
        }
        assert!(parks.borrow().is_empty(), "parked during spin/yield rungs");
        b.idle(0);
        assert_eq!(*parks.borrow(), vec![MIN_PARK_NS]);
    }

    #[test]
    fn park_period_doubles_and_caps_at_max() {
        let parks = RefCell::new(Vec::new());
        let mut b = recorded(&parks);
        for _ in 0..(MAX_SPINS + MAX_YIELDS + 13) {
            b.idle(0);
        }
        let want = vec![
            1_000, 2_000, 4_000, 8_000, 16_000, 32_000, 64_000, 128_000, 256_000, 512_000,
            1_000_000, 1_000_000, 1_000_000,
        ];
        assert_eq!(*parks.borrow(), want);
    }

    #[test]
    fn work_resets_the_ladder() {
        let parks = RefCell::new(Vec::new());
        let mut b = recorded(&parks);
        for _ in 0..(MAX_SPINS + MAX_YIELDS + 3) {
            b.idle(0);
        }
        assert_eq!(parks.borrow().len(), 3);
        b.idle(1); // work: full reset
        for _ in 0..(MAX_SPINS + MAX_YIELDS) {
            b.idle(0);
        }
        assert_eq!(parks.borrow().len(), 3, "post-reset rungs must not park");
        b.idle(0);
        assert_eq!(parks.borrow()[3], MIN_PARK_NS, "park restarts at MIN");
    }
}
